plugins/catalog-backend/src/database/types.ts

```diff
@@ -83,3 +83,82 @@ export type DatabaseLocationUpdateLogEvent = {
   created_at?: string;
   message?: string;
 };
+
+export type EntityFilter = {
+  key: string;
+  values: (string | null)[];
+};
+export type EntityFilters = EntityFilter[];
+
+/**
+ * An abstraction on top of the underlying database, wrapping the basic CRUD
+ * needs.
+ */
+export type Database = {
+  /**
+   * Runs a transaction.
+   *
+   * The callback is expected to make calls back into this class. When it
+   * completes, the transaction is closed.
+   *
+   * @param fn The callback that implements the transaction
+   */
+  transaction<T>(fn: (tx: unknown) => Promise<T>): Promise<T>;
```

Comment:
could add a type parameter to `Database` for the tx type right?

Reply:
Yes, that would work. I wanted to kinda explicitly make it opaque though, so the outside can't see it at all (essentially saying, by ways of API design, "don't you bother your head about this")

Reply:
I think that would be achieved with a type parameter as well? Some `Database<Tx = unknown>` kinda thing